packages/shopping/src/application.ts

```diff
@@ -99,4 +100,17 @@ export class ShoppingApplication extends BootMixin(
 
     this.bind(UserServiceBindings.USER_SERVICE).toClass(MyUserService);
   }
+
+  async start(): Promise<void> {
+    await super.start();
+    const oaiSpec = this.getSync(RestBindings.API_SPEC);
+    this.api(
+      Object.assign(oaiSpec, {
+        components: {
+          ...SECURITY_SCHEMA_SPEC,
+        },
+        security: SECURITY_SPEC,
+      }),
```

Comment:
I believe it should be fine to call `this.api()` just with Security parts, because the code gathering spec from controllers & routes will merge that new spec with the object supplied to `this.api()`.

Have you tried to call `this.api()` from the constructor? 

```ts
export class ShoppingApplication extends BootMixin(
  ServiceMixin(RepositoryMixin(RestApplication)),
) {
  constructor(options?: ApplicationConfig) {
    super(options);

    this.api({
      components: {
          ...SECURITY_SCHEMA_SPEC,
      },
      security: SECURITY_SPEC,
  });

  // etc.
}
```

In my opinion, the security spec does not depend on the app being started, therefore it should be available even before `app.start()` is called. Thoughts?

Reply:
Should we perhaps add a story to improve `app.api()` to intelligently merge the new spec with any spec provided by previous calls of `app.api()`?

Reply:
@bajtos Now `app.api()` takes in an entire spec in type `OpenAPISpec`, not `Partial<OpenAPISpec>`, that's why it's not possible to merge a partial spec into the generated server's spec.

Do we want to change its behaviour to take in a **partial** OpenAPI spec? If so I can create a story for it :)

Reply:
Well, what are the required fields? Isn't it just a `paths` object that can be empty?

```ts
this.api({
  components: {
    ...SECURITY_SCHEMA_SPEC,
  },
  security: SECURITY_SPEC,
  paths: {
    // will be filled by LoopBack from controller metadata
  }
});
```

Reply:
@bajtos Not really, the problem is the `openapi` and `info` fields, they are required and especially the `info` object, you cannot provide an empty value like `''` or `{}`.

See my last commit, to call `this.api()`, I have to provide valid value for `openapi` and `info`, but the rest server doesn't overwrite them later.

The good spec filled by rest server:
<img width="319" alt="Screen Shot 2019-09-16 at 11 04 53 AM" src="https://user-images.githubusercontent.com/12554153/64970041-0815dd00-d873-11e9-9dc7-889e4237997a.png">

The one generated by calling `this.api()` with "empty" `openapi` and `info` fields:

<img width="165" alt="Screen Shot 2019-09-16 at 11 08 58 AM" src="https://user-images.githubusercontent.com/12554153/64969995-f0d6ef80-d872-11e9-8118-3d5d63da63d3.png">

We can improve this by defining and checking the empty value of `openapi` and `info`, like
```ts
// in rest server
if (spec.openapi == '') {// generate openapi field }
if (spec.info.title == '' && spec.info.version == '') {// generate info field}
```

Thought? If this is the right direction for improvement I can create an issue (or just submit a PR)

Reply:
I see, thank you for explaining the details!

I think that ideally, I'd like:

- `app.api()` require fields like `info` to force the user to provide these required fields. The current implementation is good for that :+1:
- Have a way for extensions to contribute additional OpenAPI metadata. This could be a new method in `Application` or `RestServer`, but we can also use binding-based approach where the app collects OpenAPI metadata from bindings with a given tag (or other metadata-based specification). I think this is out of scope of your current work (both on the spike and on enabling auth in swagger-ui), so let's create a spike issue to research different options.

As for this spike PR, I think the current proposal (as shown in the committed code) is good enough and no further changes are necessary :+1: 

/cc @raymondfeng

Reply:
We're already exploring the idea to create an extension point to enhance the generated OpenAPI spec so that the application or other extensions can add code to transform the spec.